### PR TITLE
Update the sca policies information stored in the DB

### PR DESCRIPTION
**Pull request by Skeptor**
This PR is the development of the issue: https://github.com/wazuh/wazuh/issues/3099
This development aims to preserve the integrity of the SCA database. If any policy file changes the stored information is wiped from the DB and it is replaced with a new scan information.

### CODE REVIEW

**Comment by Skeptor** [comment]
Testing
----------
- [x] The SHA256 of each policy is saved in the table `sca_policy`
- [x] Check the entry of a policy in the table `sca_policy` is updated when the policy file changes.
   - [x] The change now is visible in the DB table
   - [x] The file has a new hash in the DB
- [ ] Change a policy outside of `/ruleset/sca`
- [ ] Run Memory Tests

File: src/analysisd/decoders/security_configuration_assessment.c [hunk 1]
```diff
@@ -380,6 +381,34 @@ static int FindPolicyInfo(Eventinfo *lf, char *policy, int *socket) {
...
+
+    snprintf(msg, OS_MAXSTR - 1, "agent %s sca query_policy_sha256 %s", lf->agent_id, policy);
+
+    if (pm_send_db(msg, response, socket) == 0) {
+        if (!strcmp(response, "ok not found")){
+            retval = 1;
+        }
+        else if (strstr(response, "err")){
+            retval = -1;
+        } else {
```
Comment (chemamartinez) [inline src/analysisd/decoders/security_configuration_assessment.c]: We are not checking here the result is the expected: `ok found <sha256>`. Why don't we follow the same structure as here? https://github.com/wazuh/wazuh/blob/99cfc0dc01bcd277b2e78b1841e6c2f768ee9616/src/analysisd/decoders/security_configuration_assessment.c#L220

File: src/analysisd/decoders/security_configuration_assessment.c [hunk 2]
```diff
@@ -926,14 +968,21 @@ static void HandleScanInfo(Eventinfo *lf,int *socket,cJSON *event) {
...
             {
                 merror("Error storing scan policy monitoring information for agent %s", lf->agent_id);
             }
-            
             break;
         default:
+            os_calloc(OS_MAXSTR, sizeof(char), old_hash);
+            if(FindPolicySHA256(lf, policy_id->valuestring, socket, old_hash) == 0){
+                if(strcmp(hash_file->valuestring, old_hash)){
+                    DeletePolicyCheck(lf, policy_id->valuestring, socket);
```
Comment (chemamartinez) [inline src/analysisd/decoders/security_configuration_assessment.c]: When deleting the policy, we are not controlling the possible errors as it is done here:
https://github.com/wazuh/wazuh/blob/99cfc0dc01bcd277b2e78b1841e6c2f768ee9616/src/analysisd/decoders/security_configuration_assessment.c#L1240

In addition, delete the policy information before its corresponding checks.

File: src/analysisd/decoders/security_configuration_assessment.c [hunk 3]
```diff
@@ -926,14 +968,21 @@ static void HandleScanInfo(Eventinfo *lf,int *socket,cJSON *event) {
...
             if (result_event < 0)
             {
                 merror("Error storing scan policy monitoring information for agent %s", lf->agent_id);
             }
-            
             break;
         default:
+            os_calloc(OS_MAXSTR, sizeof(char), old_hash);
+            if(FindPolicySHA256(lf, policy_id->valuestring, socket, old_hash) == 0){
+                if(strcmp(hash_file->valuestring, old_hash)){
```
Comment (chemamartinez) [inline src/analysisd/decoders/security_configuration_assessment.c]: Add here an informative message to tell the user that the policy file changed and the whole scan will be reported again from the agent. If not, we have no way to know what happened.

File: src/wazuh_modules/wm_sca.c [hunk 4]
```diff
@@ -2040,6 +2044,13 @@ static int wm_sca_send_summary(wm_sca_t * data, int scan_id,unsigned int passed,
         cJSON_AddStringToObject(json_summary, "hash", "error_calculating_hash");
     }
 
+    if(integrity_hash_file){
+        cJSON_AddStringToObject(json_summary, "hash_file", integrity_hash_file);
+    }
+    else {
```
Comment (chemamartinez) [inline src/wazuh_modules/wm_sca.c]: Don't include the field when it wasn't able to calculate. In addition, this case is controlled when calling the `wm_sca_send_summary` only if `integrity_hash_file` exists (line 530).

File: src/analysisd/decoders/security_configuration_assessment.c [hunk 5]
```diff
@@ -978,11 +977,21 @@ static void HandleScanInfo(Eventinfo *lf,int *socket,cJSON *event) {
...
                 if(strcmp(hash_file->valuestring, old_hash)){
-                    DeletePolicyCheck(lf, policy_id->valuestring, socket);
-                    DeletePolicy(lf, policy_id->valuestring, socket);
-                    PushDumpRequest(lf->agent_id, policy_id->valuestring, 1);
+                    int delete_status = DeletePolicy(lf, policy_id->valuestring, socket);
+                    switch (delete_status) {
+                        case 0:
+                            /* Delete checks */
+                            DeletePolicyCheck(lf, policy_id->valuestring, socket);
+                            mdebug2("Policy '%s' deleted. Requesting SCA new scan", policy_id->valuestring);
```
Comment (chemamartinez) [inline src/analysisd/decoders/security_configuration_assessment.c]: `minfo("DB content for policy '%s' outdated. Latest scan requested to the SCA module.", policy_id->valuestring);`

File: src/analysisd/decoders/security_configuration_assessment.c [hunk 6]
```diff
@@ -978,11 +977,21 @@ static void HandleScanInfo(Eventinfo *lf,int *socket,cJSON *event) {
...
+                            PushDumpRequest(lf->agent_id, policy_id->valuestring, 1);
+                            break;
+                        default:
+                            mdebug1("Error deleting policy with id '%s' from database",policy_id->valuestring);
```
Comment (chemamartinez) [inline src/analysisd/decoders/security_configuration_assessment.c]: `merror("Unable to purge DB content for policy '%s'", policy_id->valuestring);`

**Comment by Skeptor** [comment]
Memory Tests with Valgrind
----
The daemons `wazuh-db` and `ossec-analysisd` doesn't have memory leaks from this PR

image
image

In `wazuh-modulesd` there is a JSON object who remains in memory:

image

The line https://github.com/wazuh/wazuh/blob/50a0a7e759c994ddc4661d549f45b67dd82c2f64/src/wazuh_modules/wm_sca.c#L2064 creates a duplicate of the summary with recursion and this could be the problem, because the object is being deleted below.

**Comment by Skeptor** [comment]
The commit https://github.com/wazuh/wazuh/pull/3116/commits/dcaa3a95656538b4fb78c03340e477c7f21f08dd solves the memory leak detected in `wazuh-modulesd`. It changes the way to free the SCA summary from `os_free` to `cJSON_Delete`

image

**Pull request merged**